jib-core/src/main/java/com/google/cloud/tools/jib/builder/BuildSteps.java

```diff
@@ -16,23 +16,181 @@
 
 package com.google.cloud.tools.jib.builder;
 
+import com.google.cloud.tools.jib.Timer;
+import com.google.cloud.tools.jib.builder.steps.StepsRunner;
+import com.google.cloud.tools.jib.cache.Cache;
 import com.google.cloud.tools.jib.cache.CacheDirectoryNotOwnedException;
 import com.google.cloud.tools.jib.cache.CacheMetadataCorruptedException;
+import com.google.cloud.tools.jib.cache.Caches;
+import com.google.common.collect.ImmutableList;
 import java.io.IOException;
 import java.util.concurrent.ExecutionException;
 
 /** Steps for building an image. */
-public interface BuildSteps {
+public class BuildSteps {
 
-  void run()
-      throws InterruptedException, ExecutionException, CacheMetadataCorruptedException, IOException,
-          CacheDirectoryNotOwnedException;
+  /** Accepts {@link StepsRunner} by running the appropriate steps. */
+  @FunctionalInterface
+  private interface StepsRunnerConsumer {
+
+    void accept(StepsRunner stepsRunner) throws ExecutionException, InterruptedException;
+  }
+
+  private static final String DESCRIPTION_FOR_DOCKER_REGISTRY = "Building and pushing image";
```

Comment:
Should this naming be less specific (can we push to non-docker registries)?

Reply:
Currently, we only can push to Docker registries. I'm not actually sure if there are any other types of container registries out there.